File: team-app-backend/src/main/java/com/cillu/core/services/TeamService.java
package com.cillu.core.services;

import com.cillu.core.domain.Player;
import com.cillu.core.domain.Team;
import com.cillu.ports.RepositoryPort;
import com.cillu.ports.TeamServicePort;

import java.util.List;

public class TeamService implements TeamServicePort {

    private RepositoryPort repositoryPort;

    public TeamService(RepositoryPort repositoryPort){
        this.repositoryPort = repositoryPort;
    }

    @Override
    public Team createNewTeam(Team team) throws Exception {
        return repositoryPort.saveTeam(team);
    }

    @Override
    public List<Team> listTeams() throws Exception {
        return repositoryPort.getTeams();
    }

    @Override
    public Player invitePlayer(Player player) throws Exception {
        return repositoryPort.savePlayer(player);
    }

    @Override
    public List<Player> listTeamPlayers(int teamID) throws Exception {
        return repositoryPort.getPlayersByTeam(teamID);
    }
}

File: golang/core/services/team_service.go
package services

import (
	"github.com/thecillu/exagonal-architecture/golang/core/domain"
	"github.com/thecillu/exagonal-architecture/golang/ports"
)

type teamService struct {
	repository ports.RepositoryPort
}

func New(repository ports.RepositoryPort) *teamService {
	return &teamService{
		repository: repository,
	}
}

func (srv *teamService) CreateNewTeam(team domain.Team) (domain.Team, error) {
	team, err := srv.repository.SaveTeam(team)
	if err != nil {
		return domain.Team{}, err
	}
	return team, nil
}

func (srv *teamService) ListTeams() ([]domain.Team, error) {
	teams, err := srv.repository.GetTeams()
	if err != nil {
		return nil, err
	}
	return teams, nil
}

func (srv *teamService) InvitePlayer(player domain.Player) (domain.Player, error) {
	player, err := srv.repository.SavePlayer(player)
	if err != nil {
		return domain.Player{}, err
	}
	return player, nil
}

func (srv *teamService) ListTeamPlayers(team_id int64) ([]domain.Player, error) {
	players, err := srv.repository.GetPlayersByTeam(team_id)
	if err != nil {
		return nil, err
	}
	return players, nil
}
